File: Java/week11/collatz/Collatz.java

/*
 Filename: collatz.java
 Author: Mikhail Zubko
 Created: 2025/03/27
 Purpose: 
*/
import java.util.ArrayList;

public class Collatz {
  // Get the Collatz sequence as an array
  public ArrayList<Integer> getLst(int num) {
    ArrayList<Integer> lst = new ArrayList<>();
    while (num > 1) {
      if (num % 2 == 0) {
        num /= 2;
      } else {
        num = num * 3 + 1;
      }
      lst.add(num);
    }
    return lst;
  }

  // Get the length of the Collatz sequence
  public int getLen(long num) {
    int len = 0;
    while (num > 1) {
      if (num % 2 == 1) {
        num = num * 3 + 1;
      } else {
        num /= 2;
      }
      len++;
    }
    return len;
  }
}

File: Cpp/week13/collatz/collatz.cpp
#include "collatz.h"
#include <vector>

// implement the constructor and public methods
Collatz::Collatz() {}

std::vector<int> Collatz::get_lst(int num) {
  std::vector<int> collatz;
  while (num > 1) {
    if (num % 2 == 0) {
      num /= 2;
    } else {
      num = num * 3 + 1;
    }
    collatz.push_back(num);
  }
  return collatz;
}
int Collatz::get_len(long num) {
  int cnt{0};
  while (num > 1) {
    if (num % 2 == 0) {
      num /= 2;
    } else {
      num = num * 3 + 1;
    }
    cnt++;
  }
  return cnt;
}
